cloudinit/sources/DataSourceEc2.py

```diff
@@ -125,12 +127,30 @@ def _get_data(self):
                 LOG.debug("FreeBSD doesn't support running dhclient with -sf")
                 return False
             try:
-                with EphemeralDHCPv4(self.fallback_interface):
+                # ipv6 dualstack might succeed when dhcp4 fails
+                # therefore attempt dualstack call even if dhcp4 fails
+                with contextlib.ExitStack() as stack:
+                    exception = None
+                    v6_only = ""
+                    stack.enter_context(
+                        EphemeralIPv6Network(self.fallback_interface)
+                    )
+                    try:
+                        stack.enter_context(
+                            EphemeralDHCPv4(self.fallback_interface)
+                        )
+                    except NoDHCPLeaseError as e:
+                        exception = e
+                        v6_only = " using link-local ipv6"
+
                     self._crawled_metadata = util.log_time(
                         logfunc=LOG.debug,
-                        msg="Crawl of metadata service",
+                        msg="Crawl of metadata service" + v6_only,
                         func=self.crawl_metadata,
                     )
+                    if exception:
```

Comment:
I don't think we actually want to raise here. If v6 succeeded but v4 failed, this will still cause the datasource to nope out, and there's a higher level try/except that should catch if both failed.

Reply:
Agreed, thanks. Looks like the exception raise happens after the message is logged, which I think explains why the test passed.